builtin/logical/database/credentials.go

```diff
@@ -0,0 +1,172 @@
+		return random.DefaultStringGenerator.Generate(ctx, b.GetRandomReader())
+	}
+	return b.System().GeneratePasswordFromPolicy(ctx, pg.PasswordPolicy)
+}
+
+// configMap returns the configuration of the passwordGenerator
+// as a map from string to string.
+func (pg passwordGenerator) configMap() map[string]string {
+	return interfaceValuesToString(structs.Map(pg))
+}
+
+// rsaKeyGenerator generates RSA key pair credentials.
+// A zero value rsaKeyGenerator is usable.
+type rsaKeyGenerator struct {
+	// Format is the output format of the generated private key.
+	// Options include: 'pkcs8' (default)
+	Format string `mapstructure:"format" structs:"format"`
+
+	// KeyBits is the bit size of the RSA key to generate.
+	// Options include: 2048 (default), 3072, and 4096
+	KeyBits int `mapstructure:"key_bits" structs:"key_bits"`
+}
+
+// newRSAKeyGenerator returns a new rsaKeyGenerator using the given config.
+// Default values will be set on the returned rsaKeyGenerator if not provided
+// in the given config.
+func newRSAKeyGenerator(config map[string]string) (rsaKeyGenerator, error) {
+	var kg rsaKeyGenerator
+	if err := mapstructure.WeakDecode(config, &kg); err != nil {
+		return kg, err
+	}
+
+	switch strings.ToLower(kg.Format) {
+	case "":
+		kg.Format = "pkcs8"
+	case "pkcs8":
+	default:
+		return kg, fmt.Errorf("invalid format: %v", kg.Format)
+	}
+
+	switch kg.KeyBits {
+	case 0:
+		kg.KeyBits = 2048
+	case 2048, 3072, 4096:
+	default:
+		return kg, fmt.Errorf("invalid key_bits: %v", kg.KeyBits)
+	}
+
+	return kg, nil
+}
+
+// Generate generates an RSA key pair. Returns a PEM-encoded, PKIX marshaled
+// public key and a PEM-encoded private key marshaled into the configuration
+// format (in that order) or an error.
+func (kg *rsaKeyGenerator) generate(r io.Reader) ([]byte, []byte, error) {
+	reader := rand.Reader
+	if r != nil {
+		reader = r
+	}
+
+	var keyBits int
+	switch kg.KeyBits {
+	case 0:
+		keyBits = 2048
+	case 2048, 3072, 4096:
+		keyBits = kg.KeyBits
+	default:
+		return nil, nil, fmt.Errorf("invalid key_bits: %v", kg.KeyBits)
+	}
+
+	key, err := rsa.GenerateKey(reader, keyBits)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	public, err := x509.MarshalPKIXPublicKey(key.Public())
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var private []byte
+	switch strings.ToLower(kg.Format) {
+	case "", "pkcs8":
+		private, err = x509.MarshalPKCS8PrivateKey(key)
+		if err != nil {
+			return nil, nil, err
+		}
+	default:
+		return nil, nil, fmt.Errorf("invalid format: %v", kg.Format)
+	}
+
+	publicBlock := &pem.Block{
+		Type:  "PUBLIC KEY",
+		Bytes: public,
+	}
+	privateBlock := &pem.Block{
+		Type:  "PRIVATE KEY",
+		Bytes: private,
+	}
+
+	return pem.EncodeToMemory(publicBlock), pem.EncodeToMemory(privateBlock), nil
+}
+
+// configMap returns the configuration of the rsaKeyGenerator
+// as a map from string to string.
+func (kg rsaKeyGenerator) configMap() map[string]string {
+	return interfaceValuesToString(structs.Map(kg))
```

Comment:
Any particular reason we're using `structs` (which seems to be unmaintained) instead of `mapstructure`?

Reply:
We're using `mapstructure` to weakly decode `map[string]string` to the `rsaKeyGenerator` struct, but I didn't know you could go back the other way 🤔 Looks like maybe you can. Will give it a try.

I used `structs` because it solved the problem and is already used in other locations of Vault. I'd like to not use it though, so let me see what I can do. Thanks!

Reply:
I wasn't able to get the conversion from `rsaKeyGenerator` struct to `map[string]string` working using `mapstructure`.. I ended up removing the use of `structs` and going with something simple instead. That's done in 3eaa2ec. Let me know if you have any thoughts around this approach.

Reply:
:+1: definitely simpler